pkg/metrics/metrics_test.go

```diff
@@ -0,0 +1,202 @@
+snapshot_controller_operation_total_seconds_bucket{driver_name="driver2",operation_name="op2",operation_status="TerminatingFailure",le="2.5"} 1
+snapshot_controller_operation_total_seconds_bucket{driver_name="driver2",operation_name="op2",operation_status="TerminatingFailure",le="5"} 1
+snapshot_controller_operation_total_seconds_bucket{driver_name="driver2",operation_name="op2",operation_status="TerminatingFailure",le="10"} 1
+snapshot_controller_operation_total_seconds_bucket{driver_name="driver2",operation_name="op2",operation_status="TerminatingFailure",le="15"} 1
+snapshot_controller_operation_total_seconds_bucket{driver_name="driver2",operation_name="op2",operation_status="TerminatingFailure",le="30"} 1
+snapshot_controller_operation_total_seconds_bucket{driver_name="driver2",operation_name="op2",operation_status="TerminatingFailure",le="60"} 1
+snapshot_controller_operation_total_seconds_bucket{driver_name="driver2",operation_name="op2",operation_status="TerminatingFailure",le="120"} 1
+snapshot_controller_operation_total_seconds_bucket{driver_name="driver2",operation_name="op2",operation_status="TerminatingFailure",le="300"} 1
+snapshot_controller_operation_total_seconds_bucket{driver_name="driver2",operation_name="op2",operation_status="TerminatingFailure",le="600"} 1
+snapshot_controller_operation_total_seconds_bucket{driver_name="driver2",operation_name="op2",operation_status="TerminatingFailure",le="+Inf"} 1
+snapshot_controller_operation_total_seconds_sum{driver_name="driver2",operation_name="op2",operation_status="TerminatingFailure"} 0.100174856
+snapshot_controller_operation_total_seconds_count{driver_name="driver2",operation_name="op2",operation_status="TerminatingFailure"} 1
+`
+
+	reader = strings.NewReader(expected)
+	decoder = expfmt.NewDecoder(reader, format)
+	expectedMf := &cmg.MetricFamily{}
+	if err := decoder.Decode(expectedMf); err != nil {
+		t.Errorf("failed to decode expected %v", err)
+	}
+
+	maxSumDelta := 0.5 // maximum sum delta value in seconds
+	if !compareMetrics(expectedMf, mf, maxSumDelta) {
+		t.Errorf("failed testing, expected\n%s\n, got\n%s\n", expected, string(r))
+	}
+}
+
+func compareMetrics(expected, got *cmg.MetricFamily, maxDelta float64) bool {
+	if (got.Name == nil || *(got.Name) != *(expected.Name)) ||
+		(got.Type == nil || *(got.Type) != *(expected.Type)) ||
+		(got.Help == nil || *(got.Help) != *(expected.Help)) {
+		return false
+	}
+	if len(got.Metric) != len(expected.Metric) {
+		return false
+	}
+	for i := 0; i < len(got.Metric); i++ {
+		// labels should be the same
+		if !reflect.DeepEqual(got.Metric[i].Label, expected.Metric[i].Label) {
+			return false
+		}
+		// compare metric
+		gh := got.Metric[i].Histogram
+		eh := expected.Metric[i].Histogram
+		if gh == nil {
+			return false
+		}
+		if gh.SampleSum == nil || math.Abs(*(gh.SampleSum)-*(eh.SampleSum)) > maxDelta {
+			return false
+		}
+		if gh.SampleCount == nil || *(gh.SampleCount) != *(eh.SampleCount) {
```

Comment:
do we need a nil check for eh.SampleCount

Reply:
no I ignored nil check purposefully as the expected output should be parsed successfully with those information available.